README.md

```diff
@@ -1 +1,544 @@
+
+The `HTTPHeaders` type represents HTTP headers.
+
+```swift
+public typealias HTTPHeaders = [HTTPHeader]
+```
+
+### Motivation
+
+Unfortunately the HTTP protocol allows duplicate headers (see `Set-Cookie`). So we have to make it an array instead of a dictionary.
+
+## HTTPBody
+
+The `HTTPBody` type represents an HTTP body in buffer or stream form.
+
+```swift
+public enum HTTPBody {
+    case BufferBody(Data)
+    case StreamBody(Stream)
+}
+```
+
+### Motivation
+
+Buffer bodies are best suited for HTTP messages with `Content-Length` headers. Stream bodies are best suited for HTTP messages with `Transfer-Encoding` set to `chunked`.
+
+## Storage
+
+The `Storage` type represents arbitrary data that can be passed between middleware and a responder in a chain.
+
+```swift
+public typealias Storage = [String: Any]
+```
+
+### Motivation
+
+Having a storage in the HTTP messages makes middleware much more useful since they can pass forward parsed/serialized content, session information or just about any custom data.
+
+## HTTPMessage
+
+The `HTTPMessage` protocol represents properties common to HTTP messages (request/response).
+
+```swift
+public protocol HTTPMessage {
+    var version: HTTPVersion { get set }
+    var headers: HTTPHeaders { get set }
+    var body: HTTPBody { get set }
+    var storage: Storage { get set }
+}
+```
+
+### Motivation
+
+Having the `HTTPMessage` protocol makes adding computed properties common to requests and responses more convenient. It removes code duplication.
+
+## HTTPMethod
+
+The `HTTPMethod` type represents an HTTP method.
+
+```swift
+public enum HTTPMethod {
+    case Delete
+    case Get
+    case Head
+    case Post
+    case Put
+    case Connect
+    case Options
+    case Trace
+    case Copy
+    case Lock
+    case MkCol
+    case Move
+    case PropFind
+    case PropPatch
+    case Search
+    case Unlock
+    case Bind
+    case Rebind
+    case Unbind
+    case ACL
+    case Report
+    case MkActicity
+    case Checkout
+    case Merge
+    case MSearch
+    case Notify
+    case Subscribe
+    case Unsubscribe
+    case Patch
+    case Purge
+    case MkCalendar
+    case Link
+    case Unlink
+    case Raw(method: String)
+}
+```
+
+### Motivation
+
+Having `HTTPMethod` as an enum with a `Raw` case makes it convenient and flexible at the same time.
+
+## URI
+
+The `URI` type represents an URI.
+
+```swift
+public struct URI {
+    public struct UserInfo {
+        public var username: String
+        public var password: String
+    }
+
+    public var scheme: String?
+    public var userInfo: UserInfo?
+    public var host: String?
+    public var port: Int?
+    public var path: String?
+    public var query: [String: String]
+    public var fragment: String?
+}
+```
+
+### Motivation
+
+Having `URI` as a `struct` allows it to conform to protocols in extensions.
+
+## HTTPRequest
+
+The `HTTPRequest` type represents an HTTP request.
+
+```swift
+public struct HTTPRequest: HTTPMessage {
```

Comment:
> Why wouldn't Kitura's request be able to conform to S4.RequestProtocol if it depends on those?

It could depend on them but only internally. As soon as the request is passed to another module, like the server for example. it will receive `S4.HTTPRequest`, which won't depend on, `BlueSocketWriter` and `CurlInvokerDelegate. Cause `KituraRequest` (fictional) is the one that depends on them, not  `S4.HTTPRequest`.